From 89c309f2af9b90de9d06f62331ecf5a3c1252587 Mon Sep 17 00:00:00 2001
From: Patrice Chotard <patrice.chotard@foss.st.com>
Date: Fri, 13 May 2022 09:51:57 +0200
Subject: [PATCH 050/120] ioport: Add resource check helpers

Add resource_overlaps() and resource_contains() helpers.
Code copied from kernel source.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git ae90d174a5fc312ff783ade7f6c9ee303710861d ioport: Add resource check helpers]
Signed-off-by: Patrice Chotard <patrice.chotard@foss.st.com>
Change-Id: I344da9830172e67f25eb3e61958cdaa63c875ac4
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/252276
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Christophe KERELLO <christophe.kerello@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/308508
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Christophe KERELLO <christophe.kerello@st.com>
---
 include/linux/ioport.h | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/include/linux/ioport.h b/include/linux/ioport.h
index 85288c3729a..c12a7f70ad7 100644
--- a/include/linux/ioport.h
+++ b/include/linux/ioport.h
@@ -135,6 +135,22 @@ static inline unsigned long resource_type(const struct resource *res)
 	return res->flags & IORESOURCE_TYPE_BITS;
 }
 
+/* True iff r1 completely contains r2 */
+static inline bool resource_contains(struct resource *r1, struct resource *r2)
+{
+	if (resource_type(r1) != resource_type(r2))
+		return false;
+	if (r1->flags & IORESOURCE_UNSET || r2->flags & IORESOURCE_UNSET)
+		return false;
+	return r1->start <= r2->start && r1->end >= r2->end;
+}
+
+/* True if any part of r1 overlaps r2 */
+static inline bool resource_overlaps(struct resource *r1, struct resource *r2)
+{
+	return r1->start <= r2->end && r1->end >= r2->start;
+}
+
 /* Convenience shorthand with allocation */
 #define request_region(start,n,name)	__request_region(&ioport_resource, (start), (n), (name), 0)
 #define __request_mem_region(start,n,name, excl) __request_region(&iomem_resource, (start), (n), (name), excl)
-- 
2.47.2

